### ISSUE TITLE
Base path is overwriten when serving swagger UI

### ISSUE THREAD

**Issue by vnenkpet** [scottie1984/swagger-ui-express#183]
I am not sure if this is actually a problem that should be adressed here. I'm using Nestjs's SwaggerModule that uses swagger-ui-express under the hood.

My problem is that when I configure the path where to load docs, e.g. /doc, it works alright on localhost (I can view the Swagger UI on localhost:8080/docs), but after I upload the same code to something like google cloud platform where the base URL already contains some path (e.g. /app-1) and I then visit /app-1/docs, it redirects to /docs and throws error. However all the other route paths configured in Nestjs work properly without having to set anything up.

**Comment by AnirudhaGohokar** [comment]
Same issue here. works perfectly on local say `http://localhost:4040/api/docs` and when deployed on server say `https://domain.com/servicename/api/docs` it redirects to `https://domain.com/docs` and fails.

**Comment by vnenkpet** [comment]
@AnirudhaGohokar Did you find any way how to work around this? Because if not, since it's kind of critical for me, I would like to attempt to fix it myself, but not sure where to start. @scottie1984

**Comment by AnirudhaGohokar** [comment]
@vnenkpet  this may sound odd but i was able to load doc after adding `/` at end. i.e in my case i changed `https://domain.com/servicename/api/docs` to `https://domain.com/servicename/api/docs/`

**Comment by vnenkpet** [comment]
@AnirudhaGohokar Well, interesting. That actually worked for me as well. Do you use Google Cloud as well or another provider? (Just wondering if it couldn't be a G-Cloud specific problem with services routing)

**Comment by rohitkhatri** [comment]
This also fixes issue for me, but I don't want to put the trailing slash, is there a way to achieve this?

**Comment by bitreflections** [comment]
@rohitkhatri  did you find any solutions?

**Comment by rohitkhatri** [comment]
No, I haven't found any solution yet

**Comment by Rest11** [comment]
I have the same problem. It overwrites the path `https://domain.com/servicename/api/doc`s to `https://domain.com/docs`
`swagger-ui-express - 4.1.4`
The putting trailing slash at the end can resolve this issue, but it's really strange and right now I don't have another solution.

Thanks @AnirudhaGohokar for the temporary solution

**Comment by ablankenship10** [comment]
I'm having this same issue deploying to serverless, which deploys to the path `/{environmentName}/docs`, but its redirecting to just /docs which doesn't exist. Adding the trailing slash works but is not ideal.

**Comment by scottie1984** [comment]
Please review this #228 and let me know if it fixes.

**Comment by Rest11** [comment]
> Please review this #228 and let me know if it fixes.

I've cloned the code from the branch `version-5` and tested it on my remote server. Unfortunately, it hasn't been fixed and I have the same issue: `It overwrites the path https://domain.com/api/docs to https://domain.com/docs`

And again regarding temporary solution: `adding the trailing slash at the end of URL is worked`

**Comment by DevFojo** [comment]
Another thing to note is the URL called when using the try it out is the redirect URL even when using the trailing slash.

For example, if the endpoint is `/healthCheck` it would call `https://domain.com/healthCheck` instead of `https://domain.com/servicename/healthCheck`

**Comment by lucianosantana** [comment]
Apparently, the issue happens when swagger is behind a proxy and the user types the URL without the trailing slash. 

In my specific case, the proxy listens to queries under the path _/api_. So when receiving a request for _/api/doc_ (that's swagger), it strips the prefix and forwards the request to the server's path _/doc_. 

Swagger UI is normally served by _expressjs/serve-static_ if the user queries the URL with trailing slash. But if the trailing slash is missing, since the path is a directory, it redirects to the same path just adding the slash. See the [code here](https://github.com/expressjs/serve-static/blob/94feedb81682f4503ed9f8dc6d51a5c1b9bfa091/index.js#L182).

The solution should consider reading the _X-Forwarded-Prefix_ header set by the proxy (in my case Traefik) and add it to the redirect Location path. Ideally, that should be done directly on _expressjs/serve-static_. But maybe some alternative solution could be achieved by adding a middleware (something like [this proposal here](https://community.traefik.io/t/shouldnt-the-stripprefix-middleware-also-adjust-the-location-response-header/4016/2)). Finally, the last option would be to modify the Location header on the proxy, but so far that's not an option on Traefik ([link to feature request](https://github.com/traefik/traefik/issues/5809)).

**Comment by lucianosantana** [comment]
For the record, I ended up solving the proxy issue using a custom middleware:

```
export function forwardedPrefixMiddleware(req: Request, res: Response, next: Function) {
    req.originalUrl = (req.headers['x-forwarded-prefix'] || '') + req.url
    next();
}
```

**Comment by Rest11** [comment]
> Lol, I spent all day trying to find the source of this problem. I thought it was my code since I'm implementing `api-docs` on instances sitting behind an `http-express-proxy` in a pm2 cluster.
> 
> This is how I solved it
> 
> ```js
> app.use('/api-docs', (req, res, next) => {
>     if (req.originalUrl == "/api-docs") return res.redirect('api-docs/')
>     next()
> }, swaggerUi.serve, swaggerUi.setup(swCol, swOptions));
> ```
> 
> Note that the absence of the leading slash in the redirect tells express to kinda "start from the end" instead of the base path.
> 
> Hope this saves someone some time!

Thank you for your suggestion, but unfortunately, it doesn't work for me because I use nginx on my remote server and it looks for `/api/` prefix (`https://domain.com/api/docs/`). So when I remove the last `/` in the path `https://domain.com/api/docs/` it redirects me to another path without `/api/` prefix (`https://domain.com/docs`) and nginx can't proxy this request to the right service

**Comment by lucianosantana** [comment]
@Rest11 why don't you give a shot on the solution I've proposed? It's precisely solving the issue you describe. It might be the case you need to add the header "manually" on your nginx config. [Here's a reference documentation](https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/).

**Comment by Rest11** [comment]
> @Rest11 why don't you give a shot on the solution I've proposed? It's precisely solving the issue you describe. It might be the case you need to add the header "manually" on your nginx config. [Here's a reference documentation](https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/).

Thank you for your suggestion. I didn't understand which prefix I have to set in `X-Forwarded-Prefix` header in Nginx config.

For example, I have this endpoint `https://domain.com/api/v1/some-service/docs/`. Nginx checks `/api/v1/some-service` path and sends this request to the `some-service`. So when I want to see swagger docs this request `https://domain.com/api/v1/some-service/docs/` is changed to `https://domain.com/v1/some-service/docs/` (it cuts `/api/`).

So which value I have to set in `X-Forwarded-Prefix` header ?

**Comment by lucianosantana** [comment]
Have you tried "/api" ?

**Comment by Rest11** [comment]
> Have you tried "/api" ?

Sure and it doesn't work and the behavior the same.
Here is one part of my nginx config
```    # company
    location ~* ^/api/(v[1-9])/company {
proxy_set_header X-Forwarded-Prefix /api;
        proxy_set_header Host $host:$server_port;
        proxy_set_header X-NginX-Proxy true;
        proxy_ssl_session_reuse off;
        proxy_cache_bypass $http_upgrade;
        proxy_redirect off;

        rewrite ^/api/(.*)$ /$1 break;

        proxy_pass http://localhost:3002;
    }

**Comment by twosdai** [comment]
Hey I don't mean to necro this issue, however this problem does persist with NestJS and api-gateway. I am sure there is a middleware solution avaiable that could be done, but I haven't been able to find it yet. If I find anything I'll post an update here, however I did want to call it out just in case anyone stumbles on it. 

Relevant Stack Overflow Question which relates: https://stackoverflow.com/questions/70742819/nestjs-swagger-ui-generation-for-api-gateway

**Comment by fwermus** [comment]
Having same issue with google cloud on prod. Adding to a slash does not solve the problem, when and endpoint is called. Any solution for google cloud function case?

**Comment by hassanmehdi98** [comment]
I am having a similar issue when my API is connected to the AWS API Gateway on some specific path rather than the base url.
i.e. `<api-gateway-prefix>.<region>.amazonaws.com/<service-name>/`

The swagger renders perfectly fine on the configured url:
`<api-gateway-prefix>.<region>.amazonaws.com/<service-name>/api-docs`

But the API calls from the swagger goes to the base url only instead:  `<api-gateway-prefix>.<region>.amazonaws.com/<path>` which then throws the Forbidden error.

**Comment by AlexeyLukyanov** [comment]
@hassanmehdi98 tell me you found a solution to the problem?

**Comment by fwermus** [comment]
No.

**Comment by raphaelcomph-dev** [comment]
I'm using NestJS and it comes with a `addServer()` method in the DocumentBuilder, and it solved my problem. 

Context:
I had multiples Firebase Function in the same Firebase project. One of them was `backend_api_client-qa`, so the complete API url was `https://us-central1-my-project.cloudfunctions.net/backend_api_client-qa`, and it worked when I used Postman or when my frontend calls any API method. But when I tried out from swagger, it always hit `https://us-central1-my-project.cloudfunctions.net/`, without the `backend_api_client-qa`, which returns 404.

So when I was creating my `DocumentBuilder`, in the `main.ts` file, I added the `.addServer("/backend_api_client-qa")` option, and it got fixed.
